step/step.go

```diff
@@ -0,0 +1,442 @@
+package step
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	androidcache "github.com/bitrise-io/go-android/cache"
+
+	"github.com/bitrise-io/go-utils/sliceutil"
+
+	"github.com/bitrise-io/go-android/gradle"
+	"github.com/bitrise-io/go-steputils/cache"
+	"github.com/bitrise-io/go-steputils/stepconf"
+	"github.com/bitrise-io/go-steputils/tools"
+	"github.com/bitrise-io/go-utils/log"
+	"github.com/bitrise-io/go-utils/pathutil"
+	"github.com/kballard/go-shellquote"
+)
+
+// Input ...
+type Input struct {
+	ProjectLocation string `env:"project_location,dir"`
+	AppPathPattern  string `env:"app_path_pattern,required"`
+	Variant         string `env:"variant"`
+	Module          string `env:"module"`
+	BuildType       string `env:"build_type,opt[apk,aab]"`
+	Arguments       string `env:"arguments"`
+	CacheLevel      string `env:"cache_level,opt[none,only_deps,all]"`
+	DeployDir       string `env:"BITRISE_DEPLOY_DIR,dir"`
+}
+
+// Config ...
+type Config struct {
+	ProjectLocation string
+
+	Variant string
+	Module  string
+
+	AppPathPattern string
+	AppType        AppType
+	Arguments      string
+
+	CacheLevel cache.Level
+	DeployDir  string
+}
+
+type AppType string
+
+const (
+	AppTypeAPK = AppType("apk")
+	AppTypeAAB = AppType("aab")
+)
+
+var (
+	ignoredSuffixes = [...]string{"Classes", "Resources", "UnitTestClasses", "AndroidTestClasses", "AndroidTestResources"}
+)
+
+const (
+	apkEnvKey     = "BITRISE_APK_PATH"
+	apkListEnvKey = "BITRISE_APK_PATH_LIST"
+
+	aabEnvKey     = "BITRISE_AAB_PATH"
+	aabListEnvKey = "BITRISE_AAB_PATH_LIST"
+
+	mappingFileEnvKey  = "BITRISE_MAPPING_PATH"
+	mappingFilePattern = "*build/*/mapping.txt"
+)
+
+type Result struct {
+	appArtifacts []gradle.Artifact
+	mappings     []gradle.Artifact
+}
+
+type InputParser interface {
```

Comment:
I'm not sure I like this. I want to save some boilerplate too, but changing `Input` to `interface{}` just to be able to extract to go-steputils looks more painful. It would make it easier to make mistakes when converting `Input` to `Config` if there is no concrete type 😔

(Go generics can't come soon enough!)

Reply:
Go stdlib's json package has the same solution: https://pkg.go.dev/encoding/json#Unmarshal, also XML and YAML packages.

Reply:
`stepconf.Parse` already takes an `interface{}`, so we wouldn't create a "worse" API here, just make it testable. If using `stepconf.Parse` with this signature didn't lead to issues in the past, then I think we should be safe with this one as well. Wdyt?

Reply:
Refactored this to use the existing `stepconf.EnvParser`, let me know what you think. I forgot that this is already present in steputils.

Reply:
i suggest using this version: https://github.com/bitrise-io/go-steputils/pull/35

Reply:
Fixed in latest commit